### DISCUSSION TITLE
Deleting rows in pivot tables many to many

### DISCUSSION THREAD

**Discussion by hubbusysyuhada** [mikro-orm/mikro-orm#4471] [Q&A] [answered]
I have 2 tables, User and Memes which has many to many relations to map likes called `upvotes` (entities shown below)

```typescript
@orm.Entity({ tableName: "memes" })
export default class Meme {
    @orm.PrimaryKey({type: "uuid"})
    id: string = generateId();
    ...
    @orm.ManyToMany(() => User, 'memes_upvoted', { pivotTable: 'memes_upvoted_upvotes_pivot', entity: () => User })
    upvotes = new orm.Collection<User>(this);
}

@orm.Entity({ tableName: "user" })
export default class User {
    @orm.PrimaryKey({type: "uuid"})
    id: string = uuid();
    ...
    @orm.ManyToMany({ pivotTable: 'memes_upvoted_upvotes_pivot' })
    memes_upvoted = new orm.Collection<Meme>(this);
}
```

On handler, I have a method called vote to add or remove relations between those 2 entities by checking whether the connection between 1 meme and 1 user or not. If the connection is not existed, then adding a new row in the pivot table connecting those two, otherwise delete it. The problem is, I have successfully adding the designated row, but everytime I tried to delete it, the row is still there.

```typescript
  static async vote(req: FastifyRequest<{ Params: { id: string } }>, rep: FastifyReply) {
    const entityManager = await req.orm.getEm()
    const user_id = req.user.id
    const meme_id = req.params.id
    const data = await entityManager.findOne(Meme, { id: meme_id }, { populate: ['upvotes'] })

    const qb = entityManager.createQueryBuilder('memes_upvoted_upvotes_pivot')
      .select('*').where({ user_id, meme_id }).limit(1);
    const vote = await entityManager.getConnection().execute(qb.getKnexQuery())
    const user = await entityManager.findOneOrFail(User, { id: req.user.id })
    
    if (!vote[0]) {
      data.total_likes += 1
      data.upvotes.add(user) // this is to connects meme and user
    }
    else {
      data.total_likes -= 1
      data.upvotes.remove(user) // this is to disconnects meme and user
    }
    await entityManager.flush()
    rep.code(200).send({ message: `Meme ${vote[0] ? 'Down' : 'Up'}voted.` })
  }
```

I tried to console.log the data.upvotes after `data.upvotes.remove(user)` and it shows the correct value (no removed data stored). I guess the problem was in flush method not removing data from pivot table.

I could delete it manually using QueryBuilder and execute it raw, but I dont want to get the EM dirty and doing it the incorrect way. Could you please enlighten me?

note: I use version 5.7.6

**Comment by B4nan** [comment, marked as answer]
I don't see anything wrong with that snippet, could you prepare a repro (ideally a simple script, not a web app)?

Maybe it's something with the entity definition - the way you define it, the `memes_upvoted` is actually owning side, I guess you don't want that. Try it this way:

```ts
// owning side, dont specify the `mappedBy`
@orm.ManyToMany({ pivotTable: 'memes_upvoted_upvotes_pivot', entity: () => User })
upvotes = new orm.Collection<User>(this);

// inverse side, specify `mappedBy`, no need to repeat the pivot table name
@orm.ManyToMany({ mappedBy: 'upvotes', entity: () => Meme })
memes_upvoted = new orm.Collection<Meme>(this);
```

Also try upgrading to the latest, you are missing quite a lot of bug fixes.

**Reply by hubbusysyuhada** [reply]
thank you! this works just fine, so I guess it's essential to specify `mappedBy` property on the inverse side right?
